### PR TITLE
jmx surge health

### PR DESCRIPTION
**Pull request by zachary-rote**

### CODE REVIEW

File: modules/command-engine/core/src/main/scala/surge/internal/persistence/PersistentActorRegion.scala [hunk 1]
```diff
@@ -83,9 +87,29 @@ class PersistentActorRegion[M](
...
+          restartSignalPatterns = restartSignalPatterns())
+
+      registrationResult.onComplete {
+        case Failure(exception) =>
+          log.error(s"$getClass registration failed", exception)
+        case Success(_) =>
+          log.debug(s"$getClass registration succeeded")
+      }(ExecutionContext.global)
+    case Failure(error) =>
+      log.error(s"Unable to register $getClass for supervision", error)
```
Comment (sebastianharko) [inline modules/command-engine/core/src/main/scala/surge/internal/persistence/PersistentActorRegion.scala:113]: I think you have to say ${this.getClass} ???

**Review by jeffboutotte — COMMENTED** [review]
I'm about halfway through but will leave a bunch of comments now while I work through the rest

File: modules/command-engine/core/src/main/scala/surge/internal/core/SurgePartitionRouterImpl.scala [hunk 2]
```diff
@@ -97,4 +96,19 @@ private[surge] final class SurgePartitionRouterImpl(
     case Failure(error) =>
       log.error(s"Unable to register $getClass for supervision", error)
   }
+
+  private def unRegistrationCallback(): PartialFunction[Try[Ack], Unit] = {
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/command-engine/core/src/main/scala/surge/internal/core/SurgePartitionRouterImpl.scala:100]: At some point in the main branch the router was not registered any more.  I assume the starting and stopping of the router was introducing problems elsewhere??

File: modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala [hunk 3]
```diff
@@ -165,6 +170,11 @@ class KafkaPartitionShardRouterActor(
     case GetPartitionRegionAssignments => sender() ! state.partitionRegions
     case GetHealth =>
       sender() ! HealthCheck(name = "shard-router-actor", id = s"router-actor-$hashCode", status = HealthCheckStatus.UP)
+    case Shutdown =>
+      log.info("Shutting down `shared-router-actor`")
+      scheduledInitialize.cancel()
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala:175]: I don't follow.  Perhaps we can discuss this further in a call..

File: modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala [hunk 4]
```diff
@@ -173,6 +183,11 @@ class KafkaPartitionShardRouterActor(
     case msg: PartitionAssignments     => handle(state, msg)
     case msg: Terminated               => handle(state, msg)
     case GetPartitionRegionAssignments => sender() ! state.partitionRegions
+    case Shutdown =>
+      log.info("Shutting down `shared-router-actor`")
+      scheduledInitialize.cancel()
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala:188]: I don't follow. Perhaps we can discuss this further in a call..

File: modules/surge-docs/src/test/java/javadocs/commandapp/BankAccountCommandModel.java [hunk 5]
```diff
@@ -70,13 +70,13 @@
         // users can feel free to use vavr pattern matching as an alternative to instanceOf.
         if (event instanceof BankAccountCreated bankAccountCreated) {
             Optional<BankAccount> bankAccount;
-            bankAccount = Optional.of(new BankAccount(event.getAccountNumber(), bankAccountCreated.accountOwner(),
-                    bankAccountCreated.securityCode(), bankAccountCreated.balance()));
+            bankAccount = Optional.of(new BankAccount(event.getAccountNumber(), bankAccountCreated.getAccountOwner(),
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/surge-docs/src/test/java/javadocs/commandapp/BankAccountCommandModel.java:73]: done

[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/surge-docs/src/test/java/javadocs/commandapp/BankAccountCommandModel.java]: done

File: modules/common/src/test/scala/surge/internal/health/windows/actor/HealthSignalWindowActorSpec.scala [hunk 6]
```diff
@@ -69,13 +80,137 @@ class HealthSignalWindowActorSpec
...
+          windowCheckInterval = 10.milliseconds)
+      val probe = TestProbe()
+      probe.watch(actorRef.actor)
+
+      actorRef.start(Some(probe.ref))
+
+      actorRef.flush()
+
+      eventually {
+        val paused = probe.fishForMessage(max = 1.second) { case _: WindowPaused =>
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/common/src/test/scala/surge/internal/health/windows/actor/HealthSignalWindowActorSpec.scala]: done

File: modules/command-engine/core/src/test/scala/surge/internal/domain/SurgeMessagePipelineSpec.scala [hunk 7]
```diff
@@ -255,15 +255,43 @@ class SurgeMessagePipelineSpec
...
+          // wait for router-actor to be registered
+          val beforeStopRegistrations = eventually {
+            val reg = pipeline.signalBus.registrations().futureValue
+            reg.exists(p => p.componentName == "router-actor") shouldEqual true
+            reg
+          }
+
+          val acknowledgedStop: Ack = pipeline.stop().futureValue
+
+          Option(acknowledgedStop) shouldBe defined
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/command-engine/core/src/test/scala/surge/internal/domain/SurgeMessagePipelineSpec.scala]: done

File: modules/command-engine/core/src/main/scala/surge/internal/persistence/PersistentActorRegion.scala [hunk 8]
```diff
@@ -62,6 +64,8 @@ class PersistentActorRegion[M](
     kafkaProducerActor.terminate()
   }
 
+  override def restartSignalPatterns(): Seq[Pattern] = Seq(Pattern.compile("kafka.partition.lag.not.found.fatal.error"))
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/command-engine/core/src/main/scala/surge/internal/persistence/PersistentActorRegion.scala]: done

File: modules/command-engine/core/src/main/scala/surge/internal/kafka/KafkaProducerActorImpl.scala [hunk 9]
```diff
@@ -177,14 +177,16 @@ class KafkaProducerActorImpl(
         status = HealthCheckStatus.UP,
         details = Some(Map("state" -> "uninitialized")))
     case update: KTableProgressUpdate => context.become(uninitialized(lastProgressUpdate = Some(update)))
+    case ShutdownProducer             => stopPublisher()
     case _: Publish                   => stash()
     case _: IsAggregateStateCurrent   => sender().tell(false, self)
   }
 
-  private def waitingForKTableIndexing(): Receive = {
+  private def waitingForKTableIndexing(initialFenceTime: Instant): Receive = {
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/command-engine/core/src/main/scala/surge/internal/kafka/KafkaProducerActorImpl.scala]: done

File: modules/command-engine/core/src/main/scala/surge/core/KafkaProducerActor.scala [hunk 10]
```diff
@@ -145,30 +151,30 @@ class KafkaProducerActor(
...
-          log.error(s"$getClass registration failed", exception)
-        case Success(_) =>
-          log.debug(s"$getClass registration succeeded")
-      }
+        signalBus.register(
+          control = this,
+          componentName = s"kafka-producer-actor-${assignedPartition.topic()}-${assignedPartition.partition()}",
+          restartSignalPatterns = restartSignalPatterns())
+
+      Option(registrationResult).foreach(result =>
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/command-engine/core/src/main/scala/surge/core/KafkaProducerActor.scala]: done

**Review by jeffboutotte — DISMISSED** [review]

File: modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala]: done

File: modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/common/src/main/scala/surge/kafka/KafkaPartitionShardRouterActor.scala]: done

File: modules/command-engine/core/src/main/scala/surge/internal/core/SurgePartitionRouterImpl.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/command-engine/core/src/main/scala/surge/internal/core/SurgePartitionRouterImpl.scala]: done

File: modules/common/src/test/scala/surge/internal/health/windows/stream/sliding/SlidingHealthSignalStreamSpec.scala [hunk 11]
```diff
@@ -52,24 +52,24 @@ class SlidingHealthSignalStreamSpec
...
-        times = 5,
-        atomicMatcher =
-          SignalNameEqualsMatcher("test.trace", Some(SideEffect(Seq(HealthSignal("health.signal", "boom", SignalType.ERROR, Error("bah", None)))))),
-        Some(SideEffect(Seq(signal)))))
+    val signal = HealthSignal(topic = "health.signal", name = s"5 in a row", signalType = SignalType.TRACE, data = Trace(s"5 in a row"), source = None)
+
+    val definition = SignalPatternMatcherDefinition.repeating(
+      times = 5,
+      pattern = Pattern.compile("test.trace"),
+      frequency = 10.seconds,
```
[the remark this answers is not in the document]
Comment (zachary-rote) [reply modules/common/src/test/scala/surge/internal/health/windows/stream/sliding/SlidingHealthSignalStreamSpec.scala]: done

**Review by jeffboutotte — APPROVED** [review]

**Pull request merged**